### ISSUE TITLE
 Should we do away with 4S secret storage in favour of deriving everything deterministically from the recovery key?

### ISSUE THREAD

**Issue by dkasak** [matrix-org/matrix-spec#1987]
## Suggestion

I just had a mad scientist idea: why do we bother with Secret Storage (4S component, "secure storage") instead of deriving all secrets (cryptographic identity,  key, etc) deterministically from the Recovery Key (the thing with many names which we enter to access Secret Storage)?

### Pros

- Improved reliability because there is no server involvement necessary
- Spec complexity reduction
- Code and attack surface area reduction due to a removal of a complex moving-part feature that is tricky to implement correctly
- Using this approach, a WebAuthn PRF-enabled key (e.g. a Yubikey) could directly act as your entire crypto identity
- We still retain the ability to:
  - Directly gossip derived secrets between devices, via encrypted to-device. So you can still grant access to parts of your identity and storage, and this access can even be partial.
  - Rotate individual secrets, by e.g. mixing in an incrementing counter value during derivation
- The recovery key becomes an (entirely?) client-side concept, divorced from the spec since it's just an input to a KDF. This would allow for faster iteration and improve experimentation with new approaches.

### Cons

I'm listing all cons I can think of, but I'll attempt to argument why I believe they are not showstoppers.

1. Stealing a recovery key immediately grants you the ability to derive all secrets, whereas before you had to both steal the recovery key and separately steal secrets encrypted with it from the server (requiring a password or access token).
   - But you still need the password or access token to do anything useful with those secrets: steal message keys or messages, log in a new device, expose signed devices to others, etc. So it doesn't appear to materially change anything in terms of exploitation.
2. You can no longer rotate your recovery key without also rotating your identity and message key storage.
   - But how important is this use case really? Empirically, a user's primary concern is retaining *access to their messages* and nothing appears to change in that regard. Consider the possible scenarios:
     - You've lost all of your devices but you still remember the recovery key: Nothing changes.
     - You have at least one fully activated device but you've lost your recovery key: Your remaining device still has access to your message keys. The only change is you'll have to re-upload them after recovery key rotation.
     - You've lost all your devices *and* you forgot your recovery key: You've lost access to your messages under either approach.
3. The flow of holding off on creating a recovery key (such as implemented in e.g. Element X) until a user is ready to set it up no longer works.
   - However, a client can still generate and remember the recovery key in the client, and then use the exact same flow, which is to offer the user to "set up recovery" during the client's lifetime and warn them to save it just before they sign out.
     - Yes, this means the recovery key would remain cached for a longer period in the client that created it. But this doesn't materially change the current situation from a security perspective when all clients cache all secrets protected by the recovery key anyway.
   - The user could ignore all cues and destroy the initial client without saving the recovery key. This leads to the scenario described in point 2.
   - Alternatively, the client can simply choose not to use this flow and instead ask the user to set the recovery key up front.
4. The entropy of the passphrase now plays a more significant role since identical passphrases will lead to identical secrets.
   - It's never a good idea to use a low entropy recovery key anyway.
   - We can ensure uniqueness of the secrets by mixing in deterministic unique information during key derivation, such as the MXID.
   - Client can use BIP39-style passphrases as the recovery key. These are a great middle-ground that ensures both memorability and security (high entropy).

Is this insane? Am I missing something?

**Comment by richvdh** [comment]
> You can no longer rotate your key-backup key without also rotating your cross-signing keys.

oh, the original description proposes: "Rotate individual secrets, by e.g. mixing in an incrementing counter value during derivation".

But I'm not sure how that would work in practice. If I'm a client looking to generate the current correct key-backup decryption key, how do I know what counter I should use? I could base it on information that the server tells me, but that would still leave us vulnerable to replay attacks from the server. @dkasak am I missing anything here?

**Comment by dkasak** [comment]
You're right, it's not as easy. I guess I was either implicitly thinking of some more involved scheme or simply made an error.

An example of a more involved scheme could be to obtain the counter from your other verified devices, but that requires you to have some left and makes the recovery key not sufficient anymore, so it's not an actual solution. I'll update the opening post.

**Comment by richvdh** [comment]
I really like this idea. I think it works and I think it represents security improvements and (ultimately) a simplification to all implementations.

The main problem I can see is that, currently, we have an entire ecosystem using recovery keys or passphrases, which relies on secrets being stored in 4S. We will need to maintain compatibility with that system for the forseeable future. There are two aspects to this.

1. New-style clients sets up the crypto secrets, old-style clients still need access. To handle this, as well as using the generated recovery key to derive the secrets, we *also* use it to derive 4S encryption keys and store the derived secret in 4S as we do today.

2. Old-style clients set up the crypto secrets, new-style clients need access. This is likely a situation we (all current and *future* client developers) will have to preserve *forever*, unless we want to force all our current users to rotate their identities at some point in the future.

   I think the algorithm for the new-style client needs to be something like:

   1. Obtain recovery key from user
   2. Use it to derive a private master cross-signing key. If the private key matches the published public key, carry on, and generate the other secrets from the recovery key.
   3. Otherwise, try using the recovery key as a 4S encryption key, and see if matching secrets are there.

   I think this preserves all the security improvements if the secret was in fact generated from the recovery key, whilst still providing compatibility for users who did their setup on older clients.

Honestly, I'm yet to be convinced that the benefits here are worth the development cost of changing things in existing clients, but I'd be really interested in seeing at least a prototype implementation of this.
